### PR TITLE
[CR] Scrolling the recipe components section

### PR DESCRIPTION
**Pull request by DanmakuDan**
(Completely new at pull requests.)

Added a function to the crafting menu so that the components list can be scrolled one line at a time with the (default) N (Edit: it is now M) key when the mode cycle thing is focused on components only. I'm wanting to know if this code is going in the right direction.

scrolled_cookie_components

### CODE REVIEW

**Comment by KA101** [comment]
We try to use left/right to scroll lists in this sort of circumstance.  Apart from that, good idea.

**Comment by narc0tiq** [comment]
I saw the picture and thought "[N]ext what"? And how do you scroll back?

File: src/crafting.cpp [hunk 1]
```diff
@@ -767,24 +786,36 @@ const recipe *select_crafting_recipe( int &batch_size )
...
+        } else if (action == "CYCLE_NEXT_COMPONENT_LINE"){
+            if(component_scrolling_offset<0){
+                component_scrolling_offset = 0;
+            }else{
+                component_scrolling_offset = component_scrolling_offset + 1;
+            }
         } else if (action == "LEFT") {
             subtab = prev_craft_subcat( tab, subtab );
             redraw = true;
+            component_scrolling_offset = 0;
```
Comment (kevingranade) [inline src/crafting.cpp]: Is there some way to do this that doesn't require every other command to handle this variable?  I can just see this getting missed when new commands are added and it going out of sync in weird ways.

File: src/requirements.cpp [hunk 2]
```diff
@@ -243,6 +253,54 @@ int requirement_data::print_list( WINDOW *w, int ypos, int xpos, int width, nc_c
     return ypos - oldy;
 }
 
+template<typename T>
```
Comment (kevingranade) [inline src/requirements.cpp]: Why is this 2x as big as print_list?  Ideally we wouldn't have two methods for this at all, but one that can handle both use cases.

**Comment by DanmakuDan** [comment]
[N]ext was referring to scrolling the next line in the component list, where the [N] is only available when looking the components only. Only the wide screen version was saying "[N]ext line" though, which is probably still unclear.

Is it better to reuse left/right for scrolling, or [m]ode to scroll the component list after it's been used to reach the components section? I kinda expected mode to do that originally in the first place.

I previously had the scrolling variable reset only on the redraw=true check, but then it was resetting the position after doing anything but scrolling.

The print_components_scrollable function was separated from print_components since the construction menu was also referencing it, and I hadn't thought about fixing its scrolling problem.


I think I can buffer the component list printout to make the navigation better, and have [m]ode just move a screenful at a time after reaching the components list instead of using a new key. I can also apply the same thing to the construction menu so all the text will be visible there as well.

**Comment by DanmakuDan** [comment]
Made the scrolling more efficient, now pressing 'm' will rotate through the following (didn't notice earlier that I could use foldstring outside of the printing function):

scroll_cookie_components_v2_1
scroll_cookie_components_v2_2
scroll_cookie_components_v2_3
scroll_cookie_components_v2_4
scroll_cookie_components_v2_5

**Comment by Rivet-the-Zombie** [comment]
Good stuff!

**Comment by KA101** [comment]
OK, thinking this was pre-freeze.

**Comment by KA101** [comment]
Unfortunately I was also thinking it worked: pressing n doesn't do anything, I'm afraid.  :-(

**Comment by KA101** [comment]
And on a recheck, no, pressing n didn't work.  Sorry.

**Comment by DanmakuDan** [comment]
Sorry, I mentioned in my update with the extra pictures that I moved the scrolling key to M, since the recipe description already reacted by scrolling to M in a similar effect (updated the first post).

**Comment by narc0tiq** [comment]
Is it supposed to? I got the impression it was remade to just force the
[m]ode listing to page through instead of skipping past a particularly long
components section.

**Comment by DanmakuDan** [comment]
What I was looking for was scrolling the component list if it got too long. The mode key would still focus on displaying Info/Tools/Components, Tools/Components, then Components on pressing the M key. If the components list is too long, the (v more) text indicates that you could continue pressing M to see the rest of it, a page of components at a time. The text is still positioned at the top for each mode initially (pictures 1/2/3 out of 5), since someone may have used the mode key to view all other recipes at the same level.

Also, h/j/k/l/left/right/up/down/pageup/pagedown/tab/shift-tab are already used for navigating the crafting menu, I don't know what would be the next sensible navigation keys added.

**Comment by DavidKeaton** [comment]
P and N would be ones some would know.

**Comment by narc0tiq** [comment]
@DanmakuDan  Sorry, you replied at the same time I did -- my earlier comment was addressed to @KA101 and effectively restated what you said.

**Comment by KA101** [comment]
OK, paging works.  Will retest and likely merge then as this is very useful for small-window folks.

**Pull request merged**